src/AggregateFunctions/AggregateFunctionKolmogorovSmirnovTest.h

```diff
@@ -0,0 +1,323 @@
+    };
+
+    std::pair<Float64, Float64> getResult(Alternative alternative, String method)
+    {
+        ::sort(x.begin(), x.end());
+        ::sort(y.begin(), y.end());
+
+        Float64 max_s = std::numeric_limits<Float64>::min();
+        Float64 min_s = std::numeric_limits<Float64>::max();
+        Float64 now_s = 0;
+        UInt64 pos_x = 0;
+        UInt64 pos_y = 0;
+        UInt64 n1 = x.size();
+        UInt64 n2 = y.size();
+
+        const Float64 n1_d = 1. / n1;
+        const Float64 n2_d = 1. / n2;
+        const Float64 tol = 1e-7;
+
+        // reference: https://en.wikipedia.org/wiki/Kolmogorov%E2%80%93Smirnov_test
+        while (pos_x < x.size() && pos_y < y.size())
+        {
+            if (y[pos_y] - x[pos_x] >= -tol)
+            {
+                now_s += n1_d;
+                if (likely(fabs(x[pos_x] - y[pos_y]) >= tol))
+                {
+                    max_s = std::max(max_s, now_s);
+                    min_s = std::min(min_s, now_s);
+                }
+                pos_x++;
+            }
+            else
+            {
+                now_s -= n2_d;
+                if (likely(fabs(x[pos_x] - y[pos_y]) >= tol))
+                {
+                    max_s = std::max(max_s, now_s);
+                    min_s = std::min(min_s, now_s);
+                }
+                pos_y++;
+            }
+        }
+        now_s += n1_d * (x.size() - pos_x) - n2_d * (y.size() - pos_y);
+        min_s = std::min(min_s, now_s);
+        max_s = std::max(max_s, now_s);
+
+        Float64 d = 0;
+        if (alternative == Alternative::TwoSided)
+            d = std::max(std::abs(max_s), std::abs(min_s));
+        else if (alternative == Alternative::Less)
+            d = -min_s;
+        else if (alternative == Alternative::Greater)
+            d = max_s;
+
+        UInt64 g = std::__gcd(n1, n2);
+        UInt64 nx_g = n1 / g;
+        UInt64 ny_g = n2 / g;
+
+        if (method == "auto")
+            method = std::max(n1, n2) <= 10000 ? "exact" : "asymp";
+        else if (method == "exact" && nx_g >= std::numeric_limits<Int32>::max() / ny_g)
+            method = "asymp";
+
+        Float64 p_value = std::numeric_limits<Float64>::infinity();
+
+        if (method == "exact")
+        {
+            /* reference:
+             * Gunar Schröer and Dietrich Trenkler
+             * Exact and Randomization Distributions of Kolmogorov-Smirnov, Tests for Two or Three Samples
+             *
+             * and
+             *
+             * Thomas Viehmann
+             * Numerically more stable computation of the p-values for the two-sample Kolmogorov-Smirnov test
+             */
+            if (n2 > n1)
+                std::swap(n1, n2);
+
+            const Float64 f_n1 = static_cast<Float64>(n1);
+            const Float64 f_n2 = static_cast<Float64>(n2);
+            const Float64 k_q = (0.5 + floor(d * f_n2 * f_n1 - tol)) / (f_n2 * f_n1);
+            PaddedPODArray<Float64> u(n1 + 1);
```

Comment:
Sorry, can you please explain what is `k_q` and `u`, because this code and formulas doesn't look like anything from papers you mentioned. I also don't understand why `u` in one-dimensional array :(

Reply:
> Sorry, can you please explain what is `k_q` and `u`, because this code and formulas doesn't look like anything from papers you mentioned. I also don't understand why `u` in one-dimensional array :(

You could just look at the second paper(Numerically more stable computation of the p-values for the two-sample Kolmogorov-Smirnov test), this algorithm is a simplification of the algorithm in the first paper(Exact and Randomization Distributions of Kolmogorov-Smirnov, Tests for Two or Three Samples)


$$
C_{i,j} = \begin{cases} 1 \quad \text {if \|{i/n-j/m}\|>=d}  \\
0 \quad \text {if \|{i/n-j/m}\|< d and (i == 0 or j == 0)}  \\
C_{i-1,j} \frac{i}{i+j} + C_{i,j-1} \frac{j}{i+j} \quad \text {therwise} \\
\end{cases} \\
$$

`u[j]` means `c[i][j]`.
We can simplify one dim with  the outer loop i.

`k_q = (0.5 + floor(d * f_n2 * f_n1 - tol)) / (f_n2 * f_n1)` means `d`.
`-tol` because of avoiding floating point precision errors.
`floor and add 0.5 / (f_n1 * f_n2)` because (i * f_n2 - j * f_n1) approximate integers.

Reply:
$$
C_{i,j} = C_{i-1,j} \frac{i}{i+j} + C_{i,j-1} \frac{j}{i+j} 
$$

Reply:
Thanks for the explanation, unfortunately I still doubt the correctness of the "simplification".
Let's take a look at this expression: 

$$
C_{i,j} = C_{i-1,j} \frac{i}{i+j} + C_{i,j-1} \frac{j}{i+j} 
$$

So basically we iterate over `i` and then over `j`:

```
for i in range(n1):
    for j in range(n2):
        c[i][j] = ...
```
In a scope of inner loop `i == constant`. Let's find out how $j$, $j * \frac{i}{i+j}$ and $(j-1) * \frac{j}{i+j}$ relate to each other. It seems that:

1) 

$$
j > j * \frac{i}{i+j} \forall i,j > 0
$$

And cases where i == 0 or j == 0 are handled separately. 

2) 

$$
j > (j-1) * \frac{j}{i+j} \forall i,j > 0
$$

because

$$
(j-1) * \frac{j}{i+j} - j = \frac{ j^2 - j -ji -j^2 }{i+j} = -j \frac{i + 1}{i + j} < 0
$$


3) 

$$
 j * \frac{i}{i+j} > (j-1) * \frac{j}{i+j}  \forall i,j > 0
$$

because

$$
 (j-1) * \frac{j}{i+j} =  j * \frac{i}{i+j} -  \frac{j}{i+j} 
$$

and

$$
\frac{j}{i+j} > 0 \forall i, j > 0
$$

So, $j$ >  $j * \frac{i}{i+j}$ > $(j-1) * \frac{j}{i+j}$ and thus we reference to previous values when iterating over j. This is Ok for component $C_{i,j-1} \frac{j}{i+j}$, but not Ok for component $C_{i-1,j} \frac{i}{i+j}$. 

And I guess we need to use at least two arrays - current "layer" and "previous" layer. Exactly what is happening in the code listing in `Numerically more stable computation of the p-values for the
two-sample Kolmogorov-Smirnov test` paper.

Reply:
Please correct me if I'm wrong.

Reply:
I'm sorry. I don't quite understand why we need to find out $j$, $j*\frac{i}{i+j}$, $(j-1)\frac{j}{i+j}$ relate to each other.

We can basicly write this algorithm:
```cpp
for (i = 0; i < n; i++) {
  for (j = 0; j < m; j++) {
    if (abs(i/n - j/m) >= d) c[i][j] = 1;
    else if (!i or !j) c[i][j] = 0;
    else {
       c[i][j] = c[i-1][j] * i / (i+j) + c[i][j-1] * j / (i+j)
       // Now we simply to one-dim:
       // At this time. 
       // c[j] equal to c[i-1][j]. Because c[j] calc in the last outer loop i, and hasn't been changed util now.
       // c[j-1] equal to c[i][j-1]. Beacuse c[j-1] calc in this inner loop j.
       // So, c[j] = c[j] * i / (i+j) + c[i][j-1] * j / (i+j)  <==> c[i][j] = c[i-1][j] * i / (i+j) + c[i][j-1] * j / (i+j)
    }
  }
}
c[n-1][m-1] is result
```